runtime/unchecked_opt.h

```diff
@@ -0,0 +1,73 @@
+#ifndef FREE_TENSOR_UNCHECKED_OPT_H
+#define FREE_TENSOR_UNCHECKED_OPT_H
+
+#include <optional>
+
+#ifdef __CUDA_ARCH__
+#define FUNC_ATTR __attribute__((always_inline)) __host__ __device__
+#else
+#define FUNC_ATTR __attribute__((always_inline))
+#endif
+
+/**
+ * Unchecked `std::optional`-like holder
+ *
+ * It holds an object of type T, or null. Users are expected to know whether
+ * there is an object. Only when there is really an object, it can be accessed
+ *
+ * If an object is currently holded and it has an non-trivial destructor,
+ * `drop()` must be called before setting `UncheckedOpt` to another value, or
+ * destructing `UncheckedOpt`
+ *
+ * `UncheckedOpt` is useful for holding an uninitialized object which does not
+ * have a default constructor
+ */
+template <class T> class UncheckedOpt {
+    union {
```

Comment:
I thought this anonymous union field is using GCC extension, but after searching around it turns out to have been specified in C11 and C++ standards. C11 further allows anonymous struct fields, while C++ doesn't. Just a note.